File: Medium/3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring.cpp
/*
     參考HUAHUA : https://zxi.mytechroad.com/blog/string/leetcode-3-longest-substring-without-repeating-characters/
     time complexity: O(n)
     space complexity: O(128)
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int ans = 0;
        vector<int> lastidx(128,-1);
        int i = 0;
        for(int j = 0; j < n; j++){
            i = max(i, lastidx[s[j]] + 1);
            ans = max(ans, j - i + 1);
            lastidx[s[j]] = j;
        }
        return ans;
    }
};

File: Medium/3. Longest Substring Without Repeating Characters/lengthOfLongestSubstring.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int [] lastidx = new int[128];
        Arrays.fill(lastidx, -1);
        int start = 0;
        int ans = 0;
        for(int i = 0; i < s.length(); i++){
            if(lastidx[s.charAt(i)] != -1){
                start = Math.max(start, lastidx[s.charAt(i)] + 1);
            }
            lastidx[s.charAt(i)] = i;
            ans = Math.max(ans, i - start + 1);
        }
        return ans;
    }
}